http-client-spi/src/main/java/software/amazon/awssdk/http/DefaultSdkHttpFullRequest.java

```diff
@@ -18,139 +18,211 @@
+    }
+
+    private Integer standardizePort(Integer port) {
+        Validate.isTrue(port == null || port >= -1,
+                        "Port must be positive (or null/-1 to indicate no port), but was '%s'", port);
+
+        if (port != null && port == -1) {
+            return null;
+        }
+
+        return port;
+    }
+
     @Override
-    public Map<String, List<String>> getHeaders() {
-        return headers;
+    public String protocol() {
+        return protocol;
     }
 
     @Override
-    public Collection<String> getValuesForHeader(String header) {
-        return headers.getOrDefault(header, Collections.emptyList());
+    public String host() {
+        return host;
     }
 
     @Override
-    public String getResourcePath() {
-        return resourcePath;
+    public int port() {
+        return Optional.ofNullable(port).orElseGet(() -> SdkHttpUtils.standardPort(protocol()));
     }
 
     @Override
-    public Map<String, List<String>> getParameters() {
-        return queryParameters;
+    public Map<String, List<String>> headers() {
+        return headers;
+    }
+
+    @Override
+    public String encodedPath() {
+        return path;
     }
 
     @Override
-    public URI getEndpoint() {
-        return endpoint;
+    public Map<String, List<String>> rawQueryParameters() {
+        return queryParameters;
     }
 
     @Override
-    public SdkHttpMethod getHttpMethod() {
+    public SdkHttpMethod method() {
         return httpMethod;
     }
 
     @Override
-    public InputStream getContent() {
-        return content;
+    public Optional<InputStream> content() {
+        return Optional.ofNullable(content);
     }
 
     @Override
-    public Builder toBuilder() {
+    public SdkHttpFullRequest.Builder toBuilder() {
         return new Builder()
+                .protocol(protocol)
+                .host(host)
+                .port(port)
+                .encodedPath(path)
+                .rawQueryParameters(queryParameters)
+                .method(httpMethod)
                 .headers(headers)
-                .resourcePath(resourcePath)
-                .httpMethod(httpMethod)
-                .endpoint(endpoint)
-                .queryParameters(queryParameters)
                 .content(content);
     }
 
     /**
      * Builder for a {@link DefaultSdkHttpFullRequest}.
      */
     static final class Builder implements SdkHttpFullRequest.Builder {
-
-        private Map<String, List<String>> headers = new HashMap<>();
-        private String resourcePath;
-        @ReviewBeforeRelease("Do we need linked hash map here?")
+        private String protocol;
+        private String host;
+        private Integer port;
+        private String path;
         private Map<String, List<String>> queryParameters = new LinkedHashMap<>();
-        private URI endpoint;
         private SdkHttpMethod httpMethod;
+        private Map<String, List<String>> headers = new HashMap<>();
```

Comment:
Should we use TreeMap here to be consistent?

Reply:
I'd actually vote for `LinkedHashMap` so that when they do a set and then a get, the order (if relevant) is preserved. The contract on the "get" is the same value as being set. It'll be re-shuffled when the message itself is built to match the request's contract.

Reply:
Didn't grok that, could you give an example?

Reply:
Sure! The javadoc on the builder says something like builder.headers() is "the same value as the one set on builder.headers(Collection<String>) (and the other methods)".

LinkedHashMap makes it more true by ensuring the elements remain in order:
```Java
HashMap<String> headers = ...;
HashMap<String> headersFromBuilder = builder.headers(headers).headers();

assertThat(headersFromBuilder).isEqualTo(headers); 
// True with LinkedHashMap, maybe False with standard hash map / sorted map.
```